src/optix_coop_vec.cpp

```diff
@@ -0,0 +1,488 @@
+        case MatrixLayout::InferencingOptimal: return OPTIX_COOP_VEC_MATRIX_LAYOUT_INFERENCING_OPTIMAL;
+        case MatrixLayout::RowMajor: return OPTIX_COOP_VEC_MATRIX_LAYOUT_ROW_MAJOR;
+        default:
+            jitc_fail("jitc_optix_coop_vec_layout_id(): unsupported layout type!");
+    }
+}
+
+// --------------------------------------------------------------------------
+// Helper routines to generate frequently used PTX fragments
+
+static uint32_t get_reg_count(uint32_t l) {
+    if (l <= 16)
+        return 16;
+    else if (l <= 64)
+        return 64;
+    else
+        return 0;
+}
+
+static void declare_buffer(const char *name, const Variable *v) {
+    uint32_t tsize  = type_size[v->type],
+             length = v->array_length;
+
+    fmt("        .local .align $u .b8 $s[$u];\n"
+        "        .reg.u64 %$s;\n"
+        "        cvta.local.u64 %$s, $s;\n",
+        tsize, name, tsize * length,
+        name,
+        name, name);
+}
+
+static void copy_to_buffer(const char *name, const Variable *v) {
+    uint32_t tsize  = type_size[v->type],
+             length = v->array_length;
+    for (uint32_t i = 0; i < length; ++i)
+        fmt("        st.local.$b [$s+$u], %cv$u_$u;\n", v, name, tsize * i, v->reg_index, i);
+}
+
+static void copy_from_buffer(const char *name, const Variable *v) {
+    uint32_t tsize  = type_size[v->type],
+             length = v->array_length;
+
+    for (uint32_t i = 0; i < length; ++i)
+        fmt("        ld.local.$b %cv$u_$u, [$s+$u];\n", v, v->reg_index, i, name, tsize * i);
+}
+
+static void put_elem(const Variable *v, uint32_t reg_count, bool trailing_comma, bool is_return_value = false) {
+    for (uint32_t i = 0; i < reg_count; ++i) {
+        if (i < v->array_length || is_return_value)
+            fmt("%cv$u_$u, ", v->reg_index, i);
+        else
+            put("%u, ");
+    }
+    if (!trailing_comma)
+        buffer.rewind(2);
+}
+
+void jitc_optix_render_coop_vec_unpack(const Variable *v, const Variable *a0) {
+    uint32_t tsize = type_size[v->type];
+
+    put("    // coop_vec_unpack\n");
+    if (tsize == 4) {
+        fmt("    mov.b32 $v, %cv$u_$u;\n", v, a0->reg_index, (uint32_t) v->literal);
+    } else {
+        fmt("    {\n"
+            "        .reg.$b %temp;\n"
+            "        cvt.u$u.u32 %temp, %cv$u_$u;\n"
+            "        mov.$b $v, %temp;\n"
+            "    }\n",
+            v,
+            tsize*8, a0->reg_index, (uint32_t) v->literal,
+            v, v);
+    }
+}
+
+void jitc_optix_render_coop_vec_accum(const Variable *v, const Variable *target,
+                                      const Variable *value, const Variable *mask) {
+    bool use_mask = !mask->is_literal() || mask->literal != 1;
+    uint32_t length = value->array_length,
+             reg_count = get_reg_count(length);
+
+    if (use_mask)
+        fmt("    @!$v bra l_$u_done;\n", mask, v->reg_index);
```

Comment:
Would this op and `outer_product_accum` benefit from more coherence within the warp? (I would expect so at least for `outer_product_accum`).
If that's the case, would it make sense to apply the mask by zero-ing out the operand so that this lane is a no-op, instead of jumping over the code? Even if OptiX can maybe wake this thread back up, presumably it's faster to let it participate directly.